cloudify_agent/api/plugins/installer.py

```diff
@@ -135,67 +140,98 @@ def _install_managed_plugin(self,
+                raise exceptions.PluginInstallationError(
+                    'Managed plugin installation found but it is '
+                    'in a corrupted state. [{0}]'.format(managed_plugin))
+
+        fields = ['package_name',
+                  'package_version',
+                  'supported_platform',
+                  'distribution',
+                  'distribution_release']
+        description = ', '.join('{0}: {1}'.format(
+            field, managed_plugin.get(field))
+            for field in fields if managed_plugin.get(field))
+
+        if matching_existing_installation:
+            self.logger.info(
+                'Using existing installation of managed plugin: {0} [{1}]'
+                .format(managed_plugin.id, description))
+        elif (deployment_id != SYSTEM_DEPLOYMENT and
+              plugin['executor'] == 'central_deployment_agent'):
+            raise exceptions.PluginInstallationError(
+                'Central deployment agent managed plugins can only be '
+                'installed using the REST plugins API. [{0}]'
+                .format(managed_plugin))
+        else:
+            self.logger.info('Installing managed plugin: {0} [{1}]'
+                             .format(managed_plugin.id, description))
+            try:
+                self._wagon_install(plugin=managed_plugin, args=args)
+                shutil.move(tmp_plugin_dir, dst_dir)
+                with open(os.path.join(dst_dir, 'plugin.id'), 'w') as f:
+                    f.write(managed_plugin.id)
+
+                # Wait for Syncthing to sync plugin files on all managers
+                if syncthing_utils:
+                    self._wait_for_syncthing()
+                self._update_plugin_status(managed_plugin.id)
+            except Exception as e:
+                tpe, value, tb = sys.exc_info()
+                raise NonRecoverableError('Failed installing managed '
+                                          'plugin: {0} [{1}][{2}]'
+                                          .format(managed_plugin.id,
+                                                  plugin, e)), None, tb
+
+    @staticmethod
+    def _wait_for_syncthing():
+        # This file exists only when the Syncthing service is running
+        if os.path.exists(syncthing_utils.config_path):
+            while not syncthing_utils.mgmtworker_is_syncing_complete():
+                time.sleep(1)
+
+    @staticmethod
+    def _update_plugin_status(plugin_id):
+        """
+        Now that both the installation and Syncthing has ended the `installing`
+        prefix is removed from the plugin's archive_name
+        :param plugin_id:
+        :return:
+        """
+        client = get_rest_client()
+        client.plugins.finish_installation(plugin_id)
+
+    @staticmethod
+    def _is_plugin_installing(plugin_id):
+        client = get_rest_client()
+        plugin = client.plugins.get(plugin_id)
+        if 'installing-' in plugin.archive_name:
```

Comment:
I think it's better to put this logic in a shared location, so does the agent will not have to impl this again